spec/index.bs

```diff
@@ -6691,7 +6709,354 @@ An <dfn dfn>Extent3D</dfn> is a {{GPUExtent3D}}.
+    <tr class='norm'>
+        <td>{{GPUTextureFormat/r8unorm}}
+        <td>{{GPUTextureComponentType/float}}
+        <td>&checkmark;
+        <td>
+        <td>&checkmark;
+    <tr class='norm'>
+        <td>{{GPUTextureFormat/r8snorm}}
+        <td>{{GPUTextureComponentType/float}}
+        <td>
+        <td>
+        <td>&checkmark;
+    <tr class='int'>
+        <td>{{GPUTextureFormat/r8uint}}
+        <td>{{GPUTextureComponentType/uint}}
+        <td>&checkmark;
+        <td>
+        <td>
+    <tr class='int'>
+        <td>{{GPUTextureFormat/r8sint}}
+        <td>{{GPUTextureComponentType/sint}}
+        <td>&checkmark;
+        <td>
+        <td>
+    <tr class='separator'><td>16-bit formats<td  colspan=4>
+    <tr class='int'>
+        <td>{{GPUTextureFormat/r16uint}}
+        <td>{{GPUTextureComponentType/uint}}
+        <td>&checkmark;
+        <td>
+        <td>
+    <tr class='int'>
+        <td>{{GPUTextureFormat/r16sint}}
+        <td>{{GPUTextureComponentType/sint}}
+        <td>&checkmark;
+        <td>
+        <td>
+    <tr class='float'>
+        <td>{{GPUTextureFormat/r16float}}
+        <td>{{GPUTextureComponentType/float}}
+        <td>&checkmark;
+        <td>
+        <td>&checkmark;
+    <tr class='norm'>
+        <td>{{GPUTextureFormat/rg8unorm}}
+        <td>{{GPUTextureComponentType/float}}
+        <td>&checkmark;
+        <td>
+        <td>&checkmark;
+    <tr class='norm'>
+        <td>{{GPUTextureFormat/rg8snorm}}
+        <td>{{GPUTextureComponentType/float}}
+        <td>
+        <td>
+        <td>&checkmark;
+    <tr class='int'>
+        <td>{{GPUTextureFormat/rg8uint}}
+        <td>{{GPUTextureComponentType/uint}}
+        <td>&checkmark;
+        <td>
+        <td>
+    <tr class='int'>
+        <td>{{GPUTextureFormat/rg8sint}}
+        <td>{{GPUTextureComponentType/sint}}
+        <td>&checkmark;
+        <td>
+        <td>
+    <tr class='separator'><td>32-bit formats<td  colspan=4>
+    <tr class='int'>
+        <td>{{GPUTextureFormat/r32uint}}
+        <td>{{GPUTextureComponentType/uint}}
+        <td>&checkmark;
+        <td>&checkmark;
+        <td>
+    <tr class='int'>
+        <td>{{GPUTextureFormat/r32sint}}
+        <td>{{GPUTextureComponentType/sint}}
+        <td>&checkmark;
+        <td>&checkmark;
+        <td>
+    <tr class='float'>
+        <td>{{GPUTextureFormat/r32float}}
+        <td>{{GPUTextureComponentType/float}}
+        <td>&checkmark;
+        <td>&checkmark;
+        <td><!-- Metal -->
+    <tr class='int'>
+        <td>{{GPUTextureFormat/rg16uint}}
+        <td>{{GPUTextureComponentType/uint}}
+        <td>&checkmark;
+        <td>
+        <td>
+    <tr class='int'>
+        <td>{{GPUTextureFormat/rg16sint}}
+        <td>{{GPUTextureComponentType/sint}}
+        <td>&checkmark;
+        <td>
+        <td>
+    <tr class='float'>
+        <td>{{GPUTextureFormat/rg16float}}
+        <td>{{GPUTextureComponentType/float}}
+        <td>&checkmark;
+        <td>&checkmark;
```

Comment:
I think this requires "UAV Typed Load" which is optional in D3D12. Dawn currently has "false" here.

Reply:
It's not optional in the API. Here is what it says:
> Hardware support optional, the format may or may not be hardware accelerated.

So it *is* supported by the API but may require a workaround by the driver.
It is different from these two:
> Required if related optional feature is supported.
> Disallowed or not available.

Reply:
I'm still not sure that's what that means.

> To programmatically verify format support in D3D11 and D3D12, refer to [Checking hardware feature support](https://docs.microsoft.com/en-us/windows/win32/direct3ddxgi/checking-hardware-feature-support).

> This section covers how to check on Format Support for Direct3D Feature Level Hardware using API calls.
> For D3D12 use [ID3D12::CheckFeatureSupport](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12device-checkfeaturesupport).

> Gets information about the features that are supported by the current graphics driver.

Seems to imply that if an optional feature isn't supported, you can't use it at all, not that it's emulated.

Reply:
I think you are confusing "supported by the driver" versus "accelerated by hardware"

Reply:
The query that they (I think) are telling you to use to check the "Hardware support optional, the format may or may not be hardware accelerated" formats is this one. If I follow the docs then I end up at "supported by the current graphics driver". I understand the wording difference but it really doesn't seem like that was the intention.

Reply:
It very well may be the case. I'm just confused as to why D3D12 baseline would be lower than Metal's...
It should be easy for Dawn or wgpu to do a quick check and see what the D3D12 validation layers think about this.

Reply:
I'm not sure the validation layers can tell us this, unless they can emulate lower-capability devices (can they?). Or they track whether we've checked the capability before using it.

Reply:
Thankfully, it's D3D12, so we don't have to find all the obscure mobile devices. We can get the low-end NV, AMD, and Intel GPUs, and see if the validation layers complain there.

Reply:
I looked around at various documentation for a while before realizing that we have Jiawei's #513.

If I'm understanding correctly from that and D3D12 docs, UAV Typed Load is just for read-write UAVs. We don't have readwrite-storage-texture in core. For readonly-storage-texture, I believe we're supposed to translate that to SRV, not UAV, and use texel fetches.

Reply:
Which is supported for every format even in FL11_0.

Reply:
I'll be reworking the storage stuff once again, tomorrow, as per our matrix discussion.

Reply:
@damyanp @RafaelCintron can you weigh in on this? What feature level does D3D12 guarantee us (and what are the allowed storage formats)?

Reply:
The lowest FL in D3D12 is FL11_0. Here are the tables in viewable form:
https://drive.google.com/drive/folders/1rGOgzzVhThGtBX60bQ53rGIAclbPS0IT?usp=sharing